[GAPIT](https://zzlab.net/GAPIT/) æ˜¯ä¸€ä¸ªç”¨äºè¿›è¡Œ GWAS çš„ R åŒ…ï¼ŒåŒ…å«äº†ä¼—å¤šæ¨¡å‹ã€‚

## å®‰è£…ä¾èµ–
GAPIT çš„è½¯ä»¶æœ¬ä½“å°±æ˜¯ä¸€ä¸ªå•ä¸€çš„ R è„šæœ¬ï¼Œæ‰€ä»¥ç›¸å…³çš„ä¾èµ–éƒ½å¾—è‡ªå·±å®‰è£…ï¼ˆè€Œä¸” GAPIT çš„è½¯ä»¶æ–‡æ¡£ä¸­å†™çš„ä¸å…¨ï¼Œéƒ½é æŠ¥é”™æç¤ºç¼ºå“ªäº›åŒ…ğŸ˜¥ï¼‰ã€‚

```R
# Use TUNA mirror
options("repos" = c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"))

# GAPIT3 dependencies for GLM, CMLM, BLINK
install.packages("gplots")
install.packages("genetics")
install.packages("ape")
install.packages("EMMREML")
install.packages("scatterplot3d")
install.packages("BiocManager")
BiocManager::install("snpStats")
BiocManager::install("multtest")
install.packages("LDheatmap")
install.packages("plotly")
install.packages("bigmemory")
```


## Hapmap æ ¼å¼

GAPIT æ‰€éœ€çš„è¾“å…¥æ–‡ä»¶ä¸ºä¸ [[TASSEL]] é€šç”¨çš„ Hapmap æ ¼å¼ã€‚

Hapmap æ–‡ä»¶çš„å‰ 11 åˆ—ä¸º SNP çš„ä¸€äº›å…ƒæ•°æ®ã€‚å…¶ä¸­ï¼Œåªæœ‰ `rs`ã€`chrom` å’Œ `pos` åˆ—ä¸º GAPIT çœŸæ­£ä¼šä½¿ç”¨çš„åˆ—ï¼Œå…¶ä½™çš„éƒ½å¯ä»¥è®¾ä¸º `NA`ã€‚

```txt
rs#     alleles chrom   pos     strand  assembly#       center  protLSID        assayLSID       panelLSID       QCcode  B001
1151    C/A     1       1151    +       NA      NA      NA      NA      NA      NA      M       C       C       A       C
1178    G/T     1       1178    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
1203    T/C     1       1203    +       NA      NA      NA      NA      NA      NA      T       T       T       T       T
1248    G/A     1       1248    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
1282    G/A     1       1282    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
1299    T/C     1       1299    +       NA      NA      NA      NA      NA      NA      T       T       T       T       T
1306    G/A     1       1306    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
1335    G/T     1       1335    +       NA      NA      NA      NA      NA      NA      G       N       G       G       G
1792    G/A     1       1792    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
2226    G/A     1       2226    +       NA      NA      NA      NA      NA      NA      G       G       G       G       G
2599    C/T     1       2599    +       NA      NA      NA      NA      NA      NA      C       C       C       C       C
```

å¯¹äºä»ç¬¬ 12 åˆ—å¼€å§‹å¾€åçš„å„æ ·æœ¬çš„åŸºå› å‹ï¼ŒGAPIT æ”¯æŒç”¨ IUPAC å•å­—æ¯ç¼©å†™æ¥è¡¨ç¤ºäºŒå€ä½“çš„åŸºå› å‹ã€‚

| Genotype | AA  | CC  | GG  | TT  | AG  | CT  | CG  | AT  | GT  | AC  |
| -------- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Code     | A   | C   | G   | T   | R   | Y   | S   | W   | K   | M   | 

å¯¹äºç¼ºå¤±çš„åŸºå› å‹ï¼ŒåŒå­—æ¯ç”¨ `NN` è¡¨ç¤ºï¼Œå•å­—æ¯ç”¨ `N` è¡¨ç¤ºã€‚

## GAPIT
ä¸€ä¸ªå…¸å‹çš„ç”¨ GAPIT è¿›è¡Œ GWAS çš„è„šæœ¬å¦‚ä¸‹æ‰€ç¤ºã€‚

```R
# Load GAPIT functions
source("scripts/gapit_functions.txt")

# Read-in phenotype data
pheno <- read.delim("rawdata/phenotype/grain_length.txt", head=FALSE)
colnames(pheno) <- c("Taxa", "GL")

# Read-in Hapmap file
hapmap <- read.delim("resources/gwas_snp.hmp.txt", head=FALSE)

# GAPIT GLM
dir.create("results/GL")
setwd("results/GL")
myGAPIT <- GAPIT(
    Y=pheno,
    G=hapmap,
    model=c("GLM", "CMLM", "Blink"),
    PCA.total=5,
    Inter.Plot=TRUE,
    Multiple_analysis=TRUE,
    file.output=T
)
setwd("../..")

```

### ä¸€äº›å‘
1. GAPIT å¯¹è¾“å…¥æ•°æ®éå¸¸æŒ‘å‰”ï¼Œå¿…é¡»å¾—æ˜¯ r-base çš„ data.frameï¼Œä¸èƒ½æ˜¯ tidyverse çš„ tibbleï¼ŒåŠå…¶ä»–ç±»ä¼¼çš„æ•°æ®ç»“æ„ã€‚
2. åœ¨ç”¨ `read.delim` å‡½æ•°è¯»å…¥ hapmap æ–‡ä»¶æ—¶ï¼Œå¿…é¡»ä½¿ç”¨ `head=FALSE` å‚æ•°ï¼Œä½¿è¯»è¿›æ¥çš„ data.frame çš„ç¬¬ä¸€è¡Œä¸º hapmap æ–‡ä»¶çš„è¡¨å¤´ã€‚

### è°ƒè¯•
ä¸å…¶ç›´æ¥è·‘å…¨éƒ¨çš„æ•°æ®ï¼Œç­‰å¾ˆé•¿æ—¶é—´ï¼Œç„¶åæŠ¥é”™ã€‚ä¸å¦‚å…ˆç”Ÿæˆä¸€ä¸ªç¤ºä¾‹æ•°æ®è¿›è¡Œè°ƒè¯•ã€‚
```bash
# å¤åˆ¶è¡¨å¤´
head -n 1 all.hmp.txt > test.hmp.txt
# éšæœºæŒ‘é€‰ 1000 è¡Œ
shuf -n 1000 all.hmp.txt >> test.hmp.txt
# æŒ‰æŸ“è‰²ä½“ä½ç½®æ’ä¸ªåº
sort -nk 1,4 test.hmp.txt > test.sort.hmp.txt
```

### æŠ¥é”™åŠè§£å†³åŠæ³•
1. `Y is empty`ï¼šè¿™æ˜¯ç”±äº GAPIT æ²¡æ³•è¯»å–è¡¨å‹å¯¼è‡´çš„ï¼Œåœ¨è¡¨å‹æ•°æ®ä¸º tibble è€Œä¸æ˜¯ data.frame æ—¶ä¼šå‡ºç°ã€‚